File: 프로그래머스/2/87946. 피로도/피로도.cpp
#include <string>
#include <vector>
#include <iostream>

int N;
std::vector<int> v;
std::vector<bool> visited;
int max = 0;

void dfs(int n, int now, std::vector<std::vector<int>> dungeons) {
    // std::cout << n << " " << now << "\n";
    if (n > max) {
        max = n;
        
    }
    for (int i = 0; i < dungeons.size(); i++) {
        if (!visited[i]) {
            visited[i] = true;
            v.emplace_back(i);
            if (dungeons[i][0] <= now) {
                // std::cout << "add " << i << "\n";
                dfs(n+1, now-dungeons[i][1], dungeons);
            }
            v.pop_back();
            visited[i] = false;
        }
    }
}

int solution(int k, std::vector<std::vector<int>> dungeons) {
    int answer = -1;
    N = dungeons.size();
    for (int i = 0; i < dungeons.size(); i++) {
        visited.emplace_back(false);
    }
    
    dfs(0, k, dungeons);
    answer = max;
    
    return answer;
}

File: 프로그래머스/2/87946. 피로도/피로도.java
class Solution {
    int count = 0;
    public int solution(int k, int[][] dungeons) {
        int answer = 0;
        boolean[] visited = new boolean[dungeons.length];
        go(k, dungeons, visited, answer);
        return count;
    }
    
    public void go(int k, int[][] dungeons, boolean[] visited, int answer) {
        for (int i = 0; i < dungeons.length; i++) {
            if (!visited[i]) {
                if (dungeons[i][0] <= k) {
                    visited[i] = true;
                    count = Math.max(answer+1, count);
                    go(k-dungeons[i][1], dungeons, visited, answer+1);
                    visited[i] = false;
                }
            }
        }
    }
}